### DISCUSSION TITLE
Problem mocking method through another method

### DISCUSSION THREAD

**Discussion by JJourdan** [stretchr/testify#1635] [Q&A]
Hi,

I've two method that implement Server : 
func (s *Server) nrCells
func (s *Server) extractTA

extractTA is being called inside nrCells.
Since I'm testing nrCells, I would like to mock extractTA.

So I tried with the following Mock structure : 

```
type MockServer struct {
	Server
	mock.Mock
}

// Mock implementation of extractTA method.
func (m *MockServer) extractTA(taID string, nrCellList []NrCellData) []NrCellData {
	logger.Info("extractTA called with taID: %s, nrCellList: %+v", taID, nrCellList)
	args := m.Called(taID, nrCellList)
	return args.Get(0).([]NrCellData)
}
```

```
func TestNrCells(t *testing.T) {
	mockServer := new(MockServer)
	t.Run("Valid ta-id parameter", func(t *testing.T) {
		mockTAID := "123"
		mockNrCellData := []NrCellData{myData}
		mockServer.On("extractTA", mock.Anything, mock.Anything).Return(mockNrCellData)
		req, err := http.NewRequest("GET", myRequest + " [ " + mockTAID+"]", http.NoBody)
		assert.NoError(t, err)
		rr := httptest.NewRecorder()
		handler := http.HandlerFunc(mockServer.nrCells)
		handler.ServeHTTP(rr, req)

		assert.Equal(t, http.StatusOK, rr.Code)

		var response []NrCellData
		err = json.Unmarshal(rr.Body.Bytes(), &response)
		assert.NoError(t, err)

		assert.Equal(t, mockNrCellData, response)
		mockServer.AssertExpectations(t)
	})
```

This is not working, I can see extractTA logs, so it's not mocked (and causing panic because of that).
But I also tried a direct call to test the mock : 
```
mockServer.On("extractTA", mock.Anything, mock.Anything).Return(mockNrCellData)
response := mockServer.extractTA(...)
assert.Equal(t, mockNrCellData, response)
```

And this is working well.
There's something wrong when I'm calling it through nrCells.
Question : Is it possible to do what I'm trying to do ? Mock a method called from another method from the same MockObject.
If yes : what's wrong here ?
If not : do I have another to reach my goal ?

Thanks

**Comment by brackendawson** [comment]
I would say that in my opinion this highlights a problem in the code, not the test. If you need to mock one method of a class to test another method of the class I would suggest that they do not belong in the same class. Does `Server` have too many responsibilities? Consider your hexagonal architecture.

Can `extractTA` be moved to a separate class? Rather than:
```go
type Server struct{}
func (s *Server) nrCells {
    s.extractTA()
}
func (s *Server) extractTA {}
```

Consider:
```go
type TAExtractor interface{
    extractTA()
}
type Server struct{
    TAExtractor
}
func (s *Server) nrCells {
   s.extractTA()
}
```
Note that this is not inheritance, your implementation of `extractTA` will have no access to `Server`.

It's not possible to write the test you want with testify. You'd need to use [monkey patching](https://github.com/bouk/monkey), which is not safe.
